### PR TITLE
Identifier contingency list implementation

### PR DESCRIPTION
**Pull request by etiennehomer**

### CODE REVIEW

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 1]
```diff
@@ -89,12 +100,14 @@ List<ContingencyListAttributes> getContingencyLists() {
     }
 
     List<ContingencyListAttributes> getContingencyLists(List<UUID> ids) {
-        return Stream.concat(
+        return Stream.of(
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/actions/server/ContingencyListService.java:103]: Why ?

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 2]
```diff
@@ -89,12 +100,14 @@ List<ContingencyListAttributes> getContingencyLists() {
     }
 
     List<ContingencyListAttributes> getContingencyLists(List<UUID> ids) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/actions/server/ContingencyListService.java:102]: Need to modify also `getContingencyLists() `

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 3]
```diff
@@ -345,6 +371,16 @@ public Optional<FormContingencyList> createFormContingencyList(UUID sourceListId
         return Optional.empty();
     }
 
+    public Optional<IdBasedContingencyList> createIdBasedContingencyList(UUID sourceListId, UUID id) {
+        IdBasedContingencyList sourceIdBasedContingencyList = getIdBasedContingencyList(sourceListId).orElse(null);
+        if (sourceIdBasedContingencyList != null) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/actions/server/ContingencyListService.java:376]: Why we do not use `createIdBasedContingencyList` method ?

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java (hunk 1 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/actions/server/ContingencyListService.java:103]: Stream.concat() takes only 2 arguments. I need to concat 3 streams here

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java (hunk 2 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/actions/server/ContingencyListService.java:104]: Done

File: src/main/java/org/gridsuite/actions/server/ContingencyListController.java [hunk 4]
```diff
@@ -141,6 +142,39 @@ public ResponseEntity<FormContingencyList> createFormContingencyList(@RequestPar
...
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .body(contingencyList))
+                .orElse(ResponseEntity.notFound().build());
+    }
+
+    @PostMapping(value = "/identifier-contingency-lists")
+    @Operation(summary = "Create a identifier contingency list from another existing one")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The identifier contingency list have been duplicated successfully"),
+            @ApiResponse(responseCode = "404", description = "Source script contingency list not found")})
+    public ResponseEntity<IdBasedContingencyList> createIdentifierContingencyList(@RequestParam("duplicateFrom") UUID sourceListId,
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/actions/server/ContingencyListController.java]: Done

File: src/main/java/org/gridsuite/actions/server/ContingencyListController.java [hunk 5]
```diff
@@ -141,6 +142,39 @@ public ResponseEntity<FormContingencyList> createFormContingencyList(@RequestPar
             .body(service.createFormContingencyList(id, formContingencyList));
     }
 
+    @PostMapping(value = "/identifier-contingency-lists", consumes = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(summary = "Create an identifier contingency list")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The identifier contingency list have been created successfully")})
+    public ResponseEntity<IdBasedContingencyList> createIdentifierContingencyList(@RequestParam(required = false, value = "id") UUID id,
+                                                                                  @RequestBody IdentifierContingencyList identifierContingencyList) {
+        return ResponseEntity.ok()
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/actions/server/ContingencyListController.java:149]: Done. Yes it's more homogeneous

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java (hunk 3 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/actions/server/ContingencyListService.java:374]: done for the 3 duplication methods

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 6]
```diff
@@ -338,9 +364,15 @@ public FormContingencyList createFormContingencyList(UUID id, FormContingencyLis
...
         if (sourceFormContingencyList != null) {
-            FormContingencyListEntity entity = new FormContingencyListEntity(sourceFormContingencyList);
-            entity.setId(id == null ? UUID.randomUUID() : id);
-            return Optional.of(fromFormContingencyListEntity(formContingencyListRepository.save(entity)));
+            return Optional.of(createFormContingencyList(id, sourceFormContingencyList));
+        }
+        return Optional.empty();
+    }
+
+    public Optional<IdBasedContingencyList> createIdBasedContingencyList(UUID sourceListId, UUID id) {
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/actions/server/ContingencyListService.java]: Done

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 7]
```diff
@@ -338,9 +364,15 @@ public FormContingencyList createFormContingencyList(UUID id, FormContingencyLis
...
+        IdBasedContingencyList sourceIdBasedContingencyList = getIdBasedContingencyList(sourceListId).orElse(null);
+        if (sourceIdBasedContingencyList != null) {
+            return Optional.of(createIdBasedContingencyList(id, sourceIdBasedContingencyList));
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/actions/server/ContingencyListService.java]: Done for the 3 duplication methods

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 8]
```diff
@@ -356,7 +382,9 @@ public void deleteContingencyList(UUID id) {
         Objects.requireNonNull(id);
         // if there is no form contingency list by this Id, deleted count == 0
         if (formContingencyListRepository.deleteFormContingencyListEntityById(id) == 0) {
-            scriptContingencyListRepository.deleteById(id);
+            if (idBasedContingencyListRepository.deleteIdBasedContingencyListEntityById(id) == 0) {
+                scriptContingencyListRepository.deleteById(id);
```
Comment (Walid-Sahnoun) [inline src/main/java/org/gridsuite/actions/server/ContingencyListService.java:386]: you could merge this two conditions in one ?

**Review by Walid-Sahnoun — APPROVED** [review]

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**